src/mame/machine/archimedes_keyb.cpp

```diff
@@ -0,0 +1,392 @@
+	PORT_BIT(0x0008, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Numlock")        PORT_CODE(KEYCODE_NUMLOCK)    PORT_CHAR(UCHAR_MAMEKEY(NUMLOCK))
+	PORT_BIT(0x0010, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("PG UP")          PORT_CODE(KEYCODE_PGUP)       PORT_CHAR(UCHAR_MAMEKEY(PGUP))
+	PORT_BIT(0x0020, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0040, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Break")          PORT_CODE(KEYCODE_PAUSE)      PORT_CHAR(UCHAR_MAMEKEY(PAUSE))
+	PORT_BIT(0x0080, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Scroll")         PORT_CODE(KEYCODE_SCRLOCK)    PORT_CHAR(UCHAR_MAMEKEY(SCRLOCK))
+	PORT_BIT(0x0100, IP_ACTIVE_LOW, IPT_OTHER)    // Keyboard ID bit 4
+	PORT_BIT(0xfe00, IP_ACTIVE_LOW, IPT_UNUSED)
+
+	PORT_START("ROW.1")
+	PORT_BIT(0x0001, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("Esc")            PORT_CODE(KEYCODE_ESC)        PORT_CHAR(UCHAR_MAMEKEY(ESC))
+	PORT_BIT(0x0002, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("`  ~")           PORT_CODE(KEYCODE_TILDE)      PORT_CHAR('`')  PORT_CHAR('~')
+	PORT_BIT(0x0004, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("z  Z")           PORT_CODE(KEYCODE_Z)          PORT_CHAR('z')  PORT_CHAR('Z')
+	PORT_BIT(0x0008, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("TAB")            PORT_CODE(KEYCODE_TAB)        PORT_CHAR('\t')
+	PORT_BIT(0x0010, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0020, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("1  !")           PORT_CODE(KEYCODE_1)          PORT_CHAR('1')  PORT_CHAR('!')
+	PORT_BIT(0x0040, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("q  Q")           PORT_CODE(KEYCODE_Q)          PORT_CHAR('q')  PORT_CHAR('Q')
+	PORT_BIT(0x0080, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("a  A")           PORT_CODE(KEYCODE_A)          PORT_CHAR('a')  PORT_CHAR('A')
+	PORT_BIT(0x0100, IP_ACTIVE_LOW, IPT_OTHER)    // Keyboard ID bit 5
+	PORT_BIT(0xfe00, IP_ACTIVE_LOW, IPT_UNUSED)
+
+	PORT_START("ROW.2")
+	PORT_BIT(0x0001, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("CAPS")           PORT_CODE(KEYCODE_CAPSLOCK)   PORT_CHAR(UCHAR_MAMEKEY(CAPSLOCK))
+	PORT_BIT(0x0002, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("x  X")           PORT_CODE(KEYCODE_X)          PORT_CHAR('x')  PORT_CHAR('X')
+	PORT_BIT(0x0004, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("s  S")           PORT_CODE(KEYCODE_S)          PORT_CHAR('s')  PORT_CHAR('S')
+	PORT_BIT(0x0008, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("F2")             PORT_CODE(KEYCODE_F2)         PORT_CHAR(UCHAR_MAMEKEY(F2))
+	PORT_BIT(0x0010, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("F1")             PORT_CODE(KEYCODE_F1)         PORT_CHAR(UCHAR_MAMEKEY(F1))
+	PORT_BIT(0x0020, IP_ACTIVE_LOW, IPT_UNUSED)
+	PORT_BIT(0x0040, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("2  \"")          PORT_CODE(KEYCODE_2)          PORT_CHAR('2')  PORT_CHAR('@')
+	PORT_BIT(0x0080, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_NAME("w  W")           PORT_CODE(KEYCODE_W)          PORT_CHAR('w')  PORT_CHAR('W')
+	PORT_BIT(0xff00, IP_ACTIVE_LOW, IPT_UNUSED)
```

Comment:
You really shouldn’t be using `PORT_NAME` for things where the name just comes from the characters – it’s repetition that risks getting out of sync.